design/automatic-network-configuration.md

```diff
@@ -0,0 +1,738 @@
+# Automatic network configuration
+
+## Status
+
+provisional
+
+## Summary
+
+This proposal is to expand the scope of Metal³ to include an API to
+manage physical network devices.
+
+## Motivation
+
+Metal³ follows the paradigm of Kubernetes Native Infrastructure (KNI),
+which is an approach to use Kubernetes to manage underlying infrastructure.
+Managing the configuration of some physical network devices is closely
+related to managing physical hosts. As bare metal hosts are provisioned
+or later repurposed, there may be corresponding physical network changes
+that must be made, such as reconfiguring a ToR switch port. If the provisioning
+of the physical host is managed through a Kubernetes API, it would be
+convenient to be able to reconfigure related network devices using a
+similar API.
+
+## Goals
+
+- Define a Kubernetes API for configuring network device.
+- Automatically configure the network infrastructure for a host when
+  adding it to a cluster.
+- Do network configuration when deprovisioning the hosts.
+- Design a network abstraction that can represent any of the target networking
+  configuration, independently of the controller used.
+
+## Non-Goals
+
+- Implement a network controller which is aimed at being an integration with
+  an existing controller only.
+- Implement a solution for a specific underlying infrastructure.
+- Implement a solution for link aggregation in the future.
+- Implement a solution for lldp in the future.
+
+## Proposal
+
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration before provisioning a BareMetalHost. First we
+added the NetworkConfiguration operator to implement the network configuration
+of the device. Then we added some methods to CAPM3 to implement together. When
+network configuration is required, CAPM3 will process some content through
+additional methods, and then trigger the NetworkConfiguration operator to
+perform network configuration.
+
+### User Stories
+
+#### Story 1
+
+As a consumer of Metal³, I hope that after adding the machine to the cluster,
+the cluster can automatically put the machine into the corresponding network.
+
+#### Story 2
+
+As a consumer of Metal³, when a machine is deleted from the cluster, the
+corresponding network configuration is automatically deleted.
```

Comment:
An obvious (to me) potential use case not covered here is when you need to connect the Host to the provisioning network for the purposes of provisioning, but you then want to remove it from that network once it has booted for security reasons.

Reply:
yes, I will improve this story later.